### PR TITLE
Redesign "Manage field names & content" dialog

### LINKED ISSUE #6536: "Manage field names & content" logic and appearance
JabRef 5.1--2020-05-26--0d7156b
Windows 10 10.0 amd64 
Java 14.0.1

I had troubles with understanding of the logic of "Edit"-"Manage field names & content" menu today. I was trying to get rid of certain fields, but they were not disappearing, in the end I figured out that to use "Clear fields" or "Set fields" you need to tick the box "overwrite existing field values", which is quite obvious now. But "Append" and "Rename" options actually work _without_ this box (why? because old values are still there?).

I think I got confused for two reasons: 
1) hard to imagine what should be the behaviour if I clear or set field without overwriting it (?)
2) gray color of the box made me think it is inactive at first, however it seems to be a default color for all the unticked boxes.
image

I think this box is simply confusing and should be removed, if you want some safety, a warning "fields will be overwritten" may be enough (with an option "do not show me this warning again"). And even that may be too much, because a person who is going to Set or Clear fields should understand that the value will change... In any case Ctrl+z will still help if something goes wrong.

Maybe also change the color of the unticked box? I don't know if I am the only one thinking that gray box is not active and cannot be ticked.

The last is a feature request: is it possible to make some real-time suggestions in the "Field name" box?
For example if I start typing "d", to drop down a list with only fields starting with "d", i.e. "date" and "doi".
This would be also useful in other places where you select one option from a list.

### PR DESCRIPTION
**Pull request by HoussemNasri**
Fixes #6536

**Question**: Looking at other dialogs in JabRef, I noticed a pattern where actions are always at the bottom of the dialog, but I can't seem to reformulate UI logic to follow that pattern. I think one solution could be to divide the UI across 2 dialogs. One for *'Edit field value of selected references'* and another for *'Copy or move the value of one field to another'*. What are your thoughts on this?

# Automatic Field Editor
## Overview
This PR aims to update the appearance & logic of the '*Manage field names & content*' dialog. Although the problem seems trivial to solve, I had to make many decisions during the design phase, hence why I included user stories and acceptance criteria's.

### 📗 Summary

**As  a** user of JabRef 

**I want to** clear one field in a collection of bib entries automatically

------

### 📗 Summary

**As a** user of JabRef 

**I want to** change the value of one field in a collection of bib entries automatically

#### Background

**Given** *'Manage field names & content'* dialog is shown

#### Acceptance Criteria 1

**When** user enter an illegal value for the given field
**Then** show an error message

#### Acceptance Criteria 2

**When** user enter a legal value to be set for the specified field 

**And** user clicks *'set field'*

**And** *'Overwrite existing field values'* checkbox is NOT selected

**Then** set the entered value to entries with the specified field empty

#### Acceptance Criteria 3

**When** user enter a legal value to be set for the specified field 

**And** user clicks *'set field'*

**And** *'Overwrite existing field values'* checkbox is selected

**Then** set the entered value to all entries even if the specified field is not empty

------

### 📗 Summary

**As a** user of JabRef

**I want to** copy the value of one field to another field in a collection of bib entries automatically

#### Background

**Given** *'Manage field names & content'* dialog is shown

#### Acceptance Criteria 1

**When** the source and destination fields are not compatible *(they don't hold the same type of data)*
**Then** show an error message

#### Acceptance Criteria 2

**When** user attempt to copy the values of source field to destination field

**And** *'Overwrite Non empty field'* is NOT selected

**Then** only update the destination field if it's empty

------

### 📗 Summary

**As a** user of JabRef

**I want to** move/cut the value of one field to another field in a collection of bib entries automatically

#### Background

**Given** *'Manage field names & content'* dialog is shown

#### Acceptance Criteria 1

**When** the source and destination fields are not compatible *(they don't hold the same type of data)*
**Then** show an error message

#### Acceptance Criteria 2

**When** *'Overwrite Non empty field'* is NOT selected

**Then** disable the *'Move values'* button

------

### 📗 Summary

**As a** user of JabRef

**I want to** append *some value* to one field in a collection of bib entries automatically

#### Background

**Given** *'Manage field names & content'* dialog is shown

#### Acceptance Criteria 1

**When** user enter an illegal value for the given field
**Then** show an error message

#### Acceptance Criteria 2

**When** *'Overwrite Non empty field'* is NOT selected
**Then** disable *'Append value'* button

------

### 📗 Summary

**As a** user of JabRef

**I want to** swap the values of two fields in a collection of bib entries automatically

#### Background

**Given** *'Manage field names & content'* dialog is shown

#### Acceptance Criteria 1

**When** *'Overwrite Non empty field'* is NOT selected
**Then** disable *'swap values'* button

## Wireframe
Screenshot 2022-06-04 023327

## Screenshots
Screenshot 2022-06-11 181620
----
Screenshot 2022-06-11 181635
----
Screenshot 2022-06-11 181646

## 💾 Technical Information

- We can use `org.jabref.logic.integrity.FieldCheckers#getForField()` to validate a field value.

.

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
Like the idea!  I am just missing the option for renaming the field here? e.g.
In the current dialog, I can rename my existing field "title" to a field "customtitle" and in the biblatex source tab you see that `title =  { mytitle } `   got replaced with  `customtitle = { mytitle }`.
Is this missing in your UI?

Regarding the Design:
We yesterday talked about the styling of the New Entry dialog and our solution was to come with tabs for example as seen in the library properties. This has a clear separation of tasks, and it also avoids long dialogs with much vertical size. This is especially often a problem for users with a small vertical screen size.

**Comment by HoussemNasri** [comment]
> I am just missing the option for renaming the field here? e.g.
In the current dialog, I can rename my existing field "title" to a field "customtitle" and in the biblatex source tab you see that title = { mytitle } got replaced with customtitle = { mytitle }.

I think I misunderstood the rename button responsibility after reading the code comment below. But, since empty fields aren't stored then it makes perfect sense. So if I understand this correctly then renaming *'Move value'* to '*Rename field*' should solve it? 
https://github.com/JabRef/jabref/blob/082e9cf480807f6f5d6423d82d9e81e73498b646/src/main/java/org/jabref/gui/edit/MassSetFieldsDialog.java?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L102-L111

> We yesterday talked about the styling of the New Entry dialog and our solution was to come with tabs for example as seen in the library properties.

Thank you; this looks way better than having 2 dialogs or putting everything in one. However, I had a look at *'Library properties'* and it has no action buttons on the body. The only action button is *'Apply.'*. Maybe we can update the dialog's buttons on tab switch?

**Comment by Siedlerchr** [comment]
I think we need to just distinguish the semantics here, although  it's technically the same process:
1. Moving a value from A to B. e.g. the convert to biblatex moves the content of the field `journal` to `journaltitle`
2. Rename a field, for example `mendeley-tags` `to mendeley_tags` 

Both are actually the same technically.  Maybe just another button Rename Fields
Regarding the action button: You can just make it a "close button". See for example in Find unlinked files.

**Comment by HoussemNasri** [comment]
I created two action buttons: 'Save' and 'Cancel'. Clicking 'Save' will save the modifications to `UndoManager` as a CompoundEdit *(Thus users can undo them by clicking the undo arrow from the global bar)*, while clicking 'Cancel' will undo all changes *(Like you have never opened the dialog)*.

**Comment by Siedlerchr** [comment]
Please do not modify the other language properties files other than English. That is handled by crowdin and will be automatically synced once we resume the sync.
we usually switch on sync shortly before a release, so that translators can go through all Changes at once.

You also probably have a leftover system.out.print somewhere, at least unit test thinks so

**Comment by HoussemNasri** [comment]
This PR has already grown larger than I expected. Some of the requirements, such as validating entered field values against selected field using `FieldCheckers`, will be left to another PR. I'd also like to write some test cases for the view models.

File: src/main/java/org/jabref/gui/JabRefFrame.java [hunk 1]
```diff
@@ -750,8 +750,7 @@ private MenuBar createMenu() {
                 new SeparatorMenuItem(),
 
                 factory.createMenuItem(StandardActions.MANAGE_KEYWORDS, new ManageKeywordsAction(stateManager)),
-                factory.createMenuItem(StandardActions.MASS_SET_FIELDS, new MassSetFieldsAction(stateManager, dialogService, undoManager)));
-
+                factory.createMenuItem(StandardActions.AUTOMATIC_FIELD_EDITOR, new AutomaticFieldEditorAction(stateManager)));
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/JabRefFrame.java:753]: You can pass the dialogService here as well

File: src/main/java/org/jabref/gui/edit/automaticfiededitor/AutomaticFieldEditorAction.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    public AutomaticFieldEditorAction(StateManager stateManager) {
+        this.stateManager = stateManager;
+
+        this.executable.bind(needsDatabase(stateManager).and(needsEntriesSelected(stateManager)));
+    }
+
+    @Override
+    public void execute() {
+        DialogService dialogService = Injector.instantiateModelOrService(DialogService.class);
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/edit/automaticfiededitor/AutomaticFieldEditorAction.java:23]: You should pass the dialogService in the constructor as well

**Review by Siedlerchr — APPROVED** [review]
Lgtm in general, just some suggestion

**Comment by calixtus** [comment]
Thanks! Great addition to JabRef!

**Comment by calixtus** [comment]
We forgot the changelog :cry:

**Comment by ThiloteE** [comment]
Hey, just tried this. Overall I like it! I think I already found a bug though T.T

How to reproduce:

1. Add two entries

```
@Article{x1,
  a = {1},
  b = {2},
}
```

```
@Article{x2,
  b = {3},
}
```

2. Select both entries
3. Go to `two fields` tab
4. From `a`, To `b`; Press `copy value`

**Result:**

```
@Article{x1,
  a = {1},
  b = {1},
}
```

```
@Article{x2,
}
```

**Expected:**

```
@Article{x1,
  a = {1},
  b = {1},
}
```

```
@Article{x2,
  b = {3},
}
```

**Pull request merged**